src/fork.mjs

```diff
@@ -0,0 +1,77 @@
+import ensureIterable from './internal/ensure-iterable'
+import { Exchange } from './internal/queues'
+import slice from './slice'
+
+function fork (number, iterable) {
+  const iterator = ensureIterable(iterable)[Symbol.iterator]()
+
+  let iterableNumber = 0
+  let noNewIterables = false
+  const exchange = new Exchange()
+  let done = false
+  let doneValue
+
+  function fetch () {
+    const newItem = iterator.next()
+    if (newItem.done) {
+      done = true
+      doneValue = newItem.value
+    } else {
+      exchange.push(newItem.value)
+    }
+  }
+
+  function returnIterator () {
+    if (noNewIterables && iterableNumber === 0) {
+      if (typeof iterator.return === 'function') iterator.return()
+    }
+  }
+
+  function * forkGen (a) {
+    try {
+      while (true) {
+        if (!a.isEmpty()) {
+          yield a.shift()
+        } else if (done) {
+          return doneValue
+        } else {
+          fetch()
+        }
+      }
+    } finally {
+      iterableNumber--
+      returnIterator()
+    }
+  }
+
+  function * generateForks () {
+    try {
+      while (true) {
+        iterableNumber++
```

Comment:
The comment makes perfectly sense but it doesn't work the way we set up (I tried it).
If I move iterableNumber in forkGen, the input iterable is closed immediately after generating the generators, because iterableNumber === 0 and noNewIterables === true.
Basically, emitting *n* forks we must consume them or they will never close the original iterator.

Reply:
I think we might be forced to use a custom iterable class for the forked iterables.

Reply:
The problem I found is a logic one, not a syntactic one.

Reply:
Maybe I don't understand the problem. I was thinking of it as closing the source too early, but you're saying the problem is that source won't get closed at all. I still think that adding the second variable should allow a solution to be expressed.

Reply:
how would you write this function:
```
  function returnIterator () {
    if (noNewIterables && numberRunningForks === 0) {
      if (typeof iterator.return === 'function') iterator.return()
    }
  }
```

Reply:
I think I am too tired to figure this out. I'll have a look tomorrow :smile:

Reply:
Yep, no hurry.
I'll see if I can tweak it this evening.

Reply:
I figured this out (incredible what a good night sleep can do to you :smile:  )
The solution is a bit hacky, but I quite like it. Basically I am initiating the fork calling "next" and I added an extra "yield"in the "forkGenerator". The result is that I am now sure that increases and decreases are executed.